Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -63,34 +88,158 @@ theorem PosSemidef.submatrix {M : Matrix n n R} (hM : M.PosSemidef) (e : m ≃ n
   simp
 #align matrix.pos_semidef.submatrix Matrix.PosSemidef.submatrix
 
+lemma conjTranspose_mul_mul_same {A : Matrix n n R} (hA : PosSemidef A)
+    (B : Matrix n n R) : PosSemidef (Bᴴ * A * B) := by
+  constructor
+  · rw [IsHermitian, conjTranspose_mul, conjTranspose_mul, conjTranspose_conjTranspose, mul_assoc,
+      hA.1]
+  · intro x
+    simpa only [star_mulVec, dotProduct_mulVec, vecMul_vecMul] using hA.2 (mulVec B x)
+
+lemma mul_mul_conjTranspose_same {A B : Matrix n n R} (hA : PosSemidef A) :
+    PosSemidef (B * A * Bᴴ) := by
+  simpa only [conjTranspose_conjTranspose] using hA.conjTranspose_mul_mul_same Bᴴ
+
+/-- The eigenvalues of a positive semi-definite matrix are non-negative -/
+lemma eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
+    (hA : Matrix.PosSemidef A) (i : n) : 0 ≤ hA.1.eigenvalues i :=
+  (hA.re_dotProduct_nonneg _).trans_eq (hA.1.eigenvalues_eq _).symm
+
+theorem transpose {M : Matrix n n R} (hM : M.PosSemidef) : Mᵀ.PosSemidef := by
+  refine ⟨IsHermitian.transpose hM.1, fun x => ?_⟩
+  convert hM.2 (star x) using 1
+  rw [mulVec_transpose, Matrix.dotProduct_mulVec, star_star, dotProduct_comm]
+
+section sqrt
+
+variable [DecidableEq n] {A : Matrix n n 𝕜} (hA : PosSemidef A)
+
+/-- The positive semidefinite square root of a positive semidefinite matrix -/
+noncomputable def sqrt := hA.1.eigenvectorMatrix *
+  diagonal ((↑) ∘ Real.sqrt ∘ hA.1.eigenvalues) * hA.1.eigenvectorMatrixᴴ
+
+lemma posSemidef_sqrt : PosSemidef hA.sqrt := by
+  apply PosSemidef.mul_mul_conjTranspose_same
+  refine posSemidef_diagonal_iff_nonneg.mpr fun i ↦ ?_
+  rw [Function.comp_apply, IsROrC.nonneg_iff]
+  constructor
+  · simp only [IsROrC.ofReal_re]
+    exact Real.sqrt_nonneg _
+  · simp only [IsROrC.ofReal_im]
+
+lemma sq_sqrt : hA.sqrt ^ 2 = A := by
```

Comment:
This is going to be annoying to work with in the goal view and documentation, since it will show as `_.sqrt ^ 2 = A`. Either we should make `A` an explicit argument of `sqrt`  (so that it prints as `A.sqrt _`, or we should write a custom elaborator that makes it print as `(_ : PosSemiDef A).sqrt`.

Reply:
In the first case you would also need to rename to `Matrix.sqrt`, right?

Reply:
I don't think it would be a good idea to rename this to `Matrix.sqrt` -- there might be other notions of square root appropriate for other kinds of matrices (not pos-semidef). Can someone give me a hint how to write custom elaborators?

Reply:
Here's how to do the second one:
```
open Lean PrettyPrinter.Delaborator SubExpr in
@[delab app.Matrix.PosSemidef.sqrt]
def delabSqrt : Delab :=
  whenPPOption getPPNotation <|
  whenNotPPOption getPPAnalysisSkip <|
  withOptionAtCurrPos `pp.analysis.skip true do
    let e ← getExpr
    guard <| e.isAppOfArity ``Matrix.PosSemidef.sqrt 6
    let optionsPerPos ← withNaryArg 5 do
      return (← read).optionsPerPos.setBool (← getPos) `pp.proofs.withType true
    withTheReader Context ({· with optionsPerPos}) delab

open Matrix

variable {A : Matrix n n 𝕜}  (hA : PosSemidef A)

#check (id hA).sqrt
```
You might have to adjust the 5 and 6 to match the actual number of arguments / position of last argument for `sqrt`